File: codequest/2019SolutionsSG/Lockheed Solutions/C/Prob20.c
#include <stdio.h>
#include <math.h>
#include <string.h>

struct bird {
    char taxonomic_family[32];
    float length;
    float width;
    float wingspan;
    float wing_angle;
    float distance;
};


int main() {

    int num_test_cases;
    scanf("%d", &num_test_cases);

    int i;
    for (i = 0; i < num_test_cases; i++) {

        int num_known;
        int num_unknown;
        scanf("%d %d", &num_known, &num_unknown);

        struct bird known[num_known];
        struct bird unknown[num_unknown];
        int j;
        for (j = 0; j < num_known; j++) {
            scanf("%s %f %f %f %f", known[j].taxonomic_family, &known[j].length, &known[j].width, &known[j].wingspan, &known[j].wing_angle);
        }
        for (j = 0; j < num_unknown; j++) {
            scanf("%f %f %f %f", &unknown[j].length, &unknown[j].width, &unknown[j].wingspan, &unknown[j].wing_angle);
        }

        int num_accipitridae;
        int num_passeridae;
        int num_cathartidae;
        for (j = 0; j < num_unknown; j++) {
            int k;
            struct bird temp;
            for (k = 0; k < num_known; k++) {
                known[k].distance = sqrt(pow(unknown[j].length - known[k].length, 2) + pow(unknown[j].width - known[k].width, 2) + pow(unknown[j].wingspan - known[k].wingspan, 2) + pow(unknown[j].wing_angle - known[k].wing_angle, 2));
            }
            for (k = 0; k < num_known-1; k++) {
                int l;
                for (l = 0; l < num_known-1; l++) {
                    if (known[l].distance > known[l+1].distance) {
                        temp = known[l];
                        known[l] = known[l+1];
                        known[l+1] = temp;
                    }
                }
            }
            num_accipitridae = 0;
            num_passeridae = 0;
            num_cathartidae = 0;
            for (k = 0; k < num_known; k++) {
                if (strcmp(known[k].taxonomic_family, "Accipitridae") == 0) {
                    num_accipitridae += 1;
                    if (num_accipitridae == 3) {
                        printf("Accipitridae\n");
                        break;
                    }
                }
                else if (strcmp(known[k].taxonomic_family, "Passeridae") == 0) {
                    num_passeridae += 1;
                    if (num_passeridae == 3) {
                        printf("Passeridae\n");
                        break;
                    }
                }
                else {
                    num_cathartidae += 1;
                    if (num_cathartidae == 3) {
                        printf("Cathartidae\n");
                        break;
                    }
                } 
            }
        }
    }
}

File: codequest/2019Solutions/Lockheed Solutions/Java/Prob20.java

/* Template.java
 *
 * Copyright (c) 2019 Lockheed Martin (LM) as an unpublished work. 
 * All Rights Reserved.
 *
 * This file does not contain any Lockheed Martin Proprietary Information.
 */

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Scanner;

public class Prob20 {

	public static void main(String[] args) {
		try (Scanner input = new Scanner(System.in)) {
			int testCases = Integer.parseInt(input.nextLine());

			for (int testcase = 0; testcase < testCases; testcase++) {
				String[] counts = input.nextLine().split(" ");
				int known = Integer.parseInt(counts[0]);
				int unknown = Integer.parseInt(counts[1]);

				List<Bird> knownBirds = new ArrayList<>();

				for (int i = 0; i < known; i++) {
					String[] info = input.nextLine().split(" ");
					Bird bird = new Bird(info[0], Double.parseDouble(info[1]), Double.parseDouble(info[2]),
							Double.parseDouble(info[3]), Double.parseDouble(info[4]));
					knownBirds.add(bird);
				}

				for (int i = 0; i < unknown; i++) {
					List<Distance> distances = new ArrayList<>();
					String[] info = input.nextLine().split(" ");
					Bird bird = new Bird(null, Double.parseDouble(info[0]), Double.parseDouble(info[1]),
							Double.parseDouble(info[2]), Double.parseDouble(info[3]));

					for (int j = 0; j < known; j++) {
						Distance distance = knownBirds.get(j).getDistance(bird);
						distances.add(distance);
					}

					Collections.sort(distances);

					int aVotes = 0;
					int cVotes = 0;
					int pVotes = 0;
					int k = 5;
					String winner = null;
					int index = 0;

					for (; index < 5; index++) {
						Distance distance = distances.get(index);
						Bird knownBird = distance.bird;

						switch (knownBird.family) {
						case "Accipitridae":
							aVotes++;
							break;
						case "Cathartidae":
							cVotes++;
							break;
						case "Passeridae":
							pVotes++;
							break;
						}
					}

					if (aVotes >= (k / 2) + 1 && aVotes > cVotes && aVotes > pVotes) {
						winner = "Accipitridae";
					} else if (cVotes >= (k / 2) + 1 && cVotes > aVotes && cVotes > pVotes) {
						winner = "Cathartidae";
					} else if (pVotes >= (k / 2) + 1 && pVotes > cVotes && pVotes > aVotes) {
						winner = "Passeridae";
					}

					while (winner == null && index < known) {
						Distance distance = distances.get(index++);
						Bird knownBird = distance.bird;

						switch (knownBird.family) {
						case "Accipitridae":
							aVotes++;
							break;
						case "Cathartidae":
							cVotes++;
							break;
						case "Passeridae":
							pVotes++;
							break;
						}
						
						if (aVotes >= (k / 2) + 1 && aVotes > cVotes && aVotes > pVotes) {
							winner = "Accipitridae";
						} else if (cVotes >= (k / 2) + 1 && cVotes > aVotes && cVotes > pVotes) {
							winner = "Cathartidae";
						} else if (pVotes >= (k / 2) + 1 && pVotes > cVotes && pVotes > aVotes) {
							winner = "Passeridae";
						}
					}
					
					System.out.println(winner);
				}

			}
		}
	}

	private static final class Distance implements Comparable<Distance> {
		public final Bird bird;
		public final int distance;

		public Distance(Bird bird, int distance) {
			this.bird = bird;
			this.distance = distance;
		}

		@Override
		public int compareTo(Distance other) {
			if (this.distance > other.distance) {
				return 1;
			} else if (this.distance < other.distance) {
				return -1;
			}
			return 0;
		}
	}

	private static final class Bird {
		public final String family;
		public final double length;
		public final double width;
		public final double wingspan;
		public final double angle;

		public Bird(String family, double length, double width, double wingspan, double angle) {
			this.family = family;
			this.length = length;
			this.width = width;
			this.wingspan = wingspan;
			this.angle = angle;
		}

		public Distance getDistance(Bird unknown) {
			double a = this.length - unknown.length;
			double b = this.width - unknown.width;
			double c = this.wingspan - unknown.wingspan;
			double d = this.angle - unknown.angle;

			double square = (a * a) + (b * b) + (c * c) + (d * d);

			return new Distance(this, (int) (Math.sqrt(square) * 100000));
		}
	}
}
